Feat IUT-39/menu group

GitJustinLau opened the pull request:
## Description
This is the component used in the navigation bar which groups menu items together. It can include single navigation items and grouped navigation items. Grouped navigation items have a collapsed subitem group which expands on click. Clicking subNav items will navigate a page. For single nav items, clicking on it will navigate to a page
NOTE: active state does not change right now because it requires a parent to pass down the state

## Link to JIRA ticket
https://implentio.atlassian.net/jira/software/projects/IUT/boards/10?selectedIssue=IUT-39

## Applicable Screenshots
![image](https://github.com/implentio/ui-toolkit/assets/93560211/9e394fe6-be40-42dd-8321-f34211303c83)

## Testing Instructions
- Fetch and checkout my branch `git fetch && git checkout <feat/IUT-39/MenuGroup>`
- npm run storybook into the CLI
- Check the following behaviors for the MenuGroup:

1. Clicking on grouped navitems will expand a subnav group and does not trigger an action
3. clicking a subnav item will trigger an action (it will not update active state!)
4. clicking a single (does not have a chevron) navItem will trigger an action  (it will not update active state!)
5. hovering will change text and chevron to purple (hover color) from blackish (default color)
6. in the GroupActive story, the action group has purple text and its subnav item has purple text and a purple bg
7. in the singleActive story, the navitem has purple text and bg
8. when toggling the MenuGroup type ("single" | "multiple"), single will only allow one grouped item to be expanded
9. if type is multiple, you can have multiple nav groups open
10. change active state through storybook controls and see if styling matches

## Checklist:
Check what is applicable by putting an 'x' in between the []:
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have tested my code thoroughly.
- [ ] I made documentation changes (optional).

joseph-hc-jun commented:
Some feedback I have from looking at deploy preview:
1. on hover & active state, let's add in the border radius.
2. `cursor: pointer` on sub nav item is missing.
3. the right/down arrow should be aligned to the right.
note: if it looks weird spanning across the whole screen on Storybook, we could put it into a div with a set width!

joseph-hc-jun commented on src/components/NavItem/NavItem.tsx:
```diff
@@ -1,34 +1,47 @@
 
-export type NavItemProps = React.HTMLAttributes<HTMLDivElement> & {
-  type: "single" | "group";
+export type singleProps = {
```
If the types are internal purposes, we might not want to export them :)

GitJustinLau replied:
oh i need to import them into the MenuGroup component. I won't include it into the src index export

joseph-hc-jun replied:
Oh that makes sense, thanks for the answer!

joseph-hc-jun commented on src/components/MenuGroup/MenuGroup.tsx:
```diff
@@ -0,0 +1,71 @@
+};
+
+const MenuGroup = ({ type, itemArr, collapsible }: accordionProps) => {
+  const accordionItems = itemArr.map((item) => {
```
One thing I'd like to point out here, this might not be a big deal in this case (and it's very commonly done too) but generally I don't think it's a good idea to define this kind of complex logic inside a component.

This way, whenever component re-renders (this happens quite a lot) this variable definition will re-run too (it's long and includes a loop, so quite expensive) which isn't necessary.

What we can do instead is,
1. defining it as a function that takes in an argument (in this case `itemArr`), and returns jsx. This way we can take the definition outside of the component and have it defined somewhere like line 20. Then you simply call it with the argument in the same place you used it.
2. You can memoize the function definition with the use of `useMemo` hook (`useCallback` if you want to make it into function). You can read more about it (here)[https://react.dev/reference/react/useMemo].

joseph-hc-jun reviewed (changes requested):
Looks great in general, requested some changes so please take a look. Thank you!

GitJustinLau commented:
> Some feedback I have from looking at deploy preview:
> 
> 1. on hover & active state, let's add in the border radius.
> 2. `cursor: pointer` on sub nav item is missing.
> 3. the right/down arrow should be aligned to the right.
>    note: if it looks weird spanning across the whole screen on Storybook, we could put it into a div with a set width!

1. done. added to default and it show stay for both hover and active. default doesnt have a bg so it wont be visible
2. added. also added to navItem (neither mentioned cursor pointer so ill put onto both)
3. it is currently set to a fixed width but not right aligned. Added right align for chevron. Removed fixed width for production. I guess it will look a bit weird but with context, it makes sense

lmk if you want any of these points adjusted

joseph-hc-jun commented on src/components/MenuGroup/MenuGroup.tsx:
```diff
@@ -0,0 +1,82 @@
+};
+
+const MenuGroup = ({ type, itemArr, collapsible }: MenuGroupProps) => {
+  const accordionItems = React.useMemo(
```
Thanks for making the change here! By defining it outside the component it would have the same effect as using `useMemo` already, so doing it here isn't strictly necessary here but it's just my small nit 😄

GitJustinLau replied:
oh i see first time using it

joseph-hc-jun reviewed (approved):
Looks great! Awesome job completing one of the more complex components! 👍
